chore: ci3 phase 1

charlielye opened the pull request:
CI3 is a conceptual goal for uniting the CI flow and the dev flow as much as possible, adding more depth to the bootstrap and build scripts to be able to handle our needs.

This PR introduces all the work on CI3 so far, but still has an earthly caller shell to make sure we can minimize the number of variables that have changed at once.

There is a lot of changes in this PR.
See https://github.com/AztecProtocol/aztec-packages/pull/10711/files for a subset of the changes without yarn.lock etc noise.

The big picture:
- The CI build has been made much less stateful. ci.yml now uses the ci3
bootstrap pattern, without fully moving off the earthly targets just
yet.
- The S3 cache mechanism is now the main cache mechanism. Note there is
no persistent disk now supporting the build.
There is a global cache on S3, readable without auth, that caches them
for 10 days. We no longer think of the build in terms of docker/buildkit
layers but instead as chunks that have different rebuild patterns that
match files in the monorepo.
- Moving to yarn 4.5.2. 

Niceties:
- faster builds due to script improvements and distributed cache
uploading by default
- work is more properly isolated in chunks from the above effort
- spot recovery is implemented, retrying with on-demand
- we no longer use github runners, side-stepping lots of edge-cases, and
instead rely on our builder realizing there is no work to do / hitting a
timeout via shutdown -P
- Docker images are no longer copied from the builder, meaning a large
class of flake is gone.

Non-niceties:
- The earthly setup is much less granular. There is two stages that have
their own one-layer builds. The earthly cache is fairly redundant, using
the S3 cache for most meaningful caching. (earthly will not be used in
ci.yml in the future)
- Some CI files are now duplicated, we will do a follow-on pass to get
rid of earthly helpers, build-system, etc
- CI currently also downloads the CI image fresh each time, will change
- we are currently pushing images to dockerhub with no expiration,
should move to ECR
- noir-projects currently retries once in the Earthfile as a last minute
issue was hit, will be fixed in a follow-up
- Issues: https://github.com/AztecProtocol/aztec-packages/issues/10775

---- 
WORKFLOW AFTER THIS PR:
- Run ./bootstrap.sh in root to bootstrap using the (publicly available) S3 cache, ./bootstrap.sh
full to force a full build
- Run ./bootstrap.sh ci to test the in-progress 'full CI3' locally
- Run ./ci.sh ec2 to test the in-progress 'full CI3' on an isolated runner
- For ci2.5, use earthly +ci in root to simulate ci.yml. This shares the S3 cache. Make sure to alias earthly to scripts/earthly_local.
- Put ci3 in your cache and note the ways to interact with ci in that
folder
- Recommended workflow, as commits are now needed to run earthly or bootstrap_ec2: 
```
# in repo root
./ci.sh draft && git commit -am "blobs work" && git push && earthly +ci
```
Other useful ci.sh commands are gha-url to see the last github job associated with your branch.
- Notable commands in ./bootstrap.sh are test-kind-network, test-e2e, images-e2e
- In yarn-project to run a single e2e test now use `test:e2e`. `test`
just runs the unit tests as per other projects.

ludamad commented on .github/workflows/ci3.yml:
```diff
@@ -0,0 +1,21 @@
+on:
+  push:
+    branches:
+      - main
```
main => master :)

ludamad commented:
nag about not gleaning anything from commit messages

ludamad commented:
If we do want this just in one command we can organize things with 
```
echo "::group::Foo"
echo "Foo"
echo "::endgroup"

echo "::group::Bar"
echo "Bar"
echo "::endgroup"
```

ludamad commented:
as for exit code 137 popping up, we can get this down but this will be tricky if we do want a shared runner, need to do memory accounting, hence why we did isolated runners

charlielye committed:
wip

charlielye committed:
wip

ludamad committed:
chore: ci3 fixes (#10280)

Co-authored-by: Charlie Lye <karl.lye@gmail.com>

charlielye committed:
wip

charlielye committed:
acir test cache flag

charlielye committed:
spelling

charlielye committed:
fix

charlielye committed:
wip

charlielye committed:
wip

charlielye committed:
hack around nargo race condition on dep dload

charlielye committed:
usage

charlielye committed:
new bootstrap for noir-contracts - super parallel

charlielye committed:
script for proto circuits

charlielye committed:
clean

charlielye committed:
noir-projects performance. full run 2m30s, cached run 13s

charlielye committed:
formatting fixes

charlielye committed:
wrong about times. 8m20s full. public_base_rollup is slow.

charlielye committed:
build sim circuits.

charlielye committed:
Always upload, if file doesn't already exist. noir-projects supports cmds.

charlielye committed:
oops

charlielye committed:
fix

charlielye committed:
fix. skip avm test

charlielye committed:
fix

charlielye committed:
normalise yp test commands as unit tests. add test:e2e and test:compose to call script. fix boxes

charlielye committed:
bootstrap_source script

charlielye committed:
fix

charlielye committed:
fix

ludamad committed:
chore: ci2.5 => ci3 merge (#10349)

- build images using new bootstrap
- scripts pass
- start of earthfile
- get rid of is_test

charlielye committed:
flatting ci3 paths. boxes test webkit and chrome.

charlielye committed:
bootstrap scripts handle cmds

charlielye committed:
fix

charlielye committed:
fix flag stdout. colors to source. local and casing.

charlielye committed:
tweaks. fixes. new ami. can grind e2e.

charlielye committed:
denoise grind fix

charlielye committed:
bring back browser test

charlielye committed:
browser is still a flake. dial back logging and polling from insane, to way-too-much.

charlielye committed:
up_quick_start test cleanup

charlielye committed:
expand and simplify acir tests.

charlielye committed:
denoiseable

charlielye committed:
acir sol tests

charlielye committed:
Merge branch 'cl/ci3-fake-base' into cl/ci3

charlielye committed:
merge fix

charlielye committed:
enable pipefail. source colors. color headers.

charlielye committed:
better call pattern. color tweak

charlielye committed:
Merge branch 'cl/ci3-fake-base' into cl/ci3

charlielye committed:
fix

charlielye committed:
crs path fix in bb.js. tweaks to allow playing with omp again (nope). limit acir test resource due to master merge blowup.

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
fix compose log level. ec2-full cmd.

charlielye committed:
fix

charlielye committed:
fix noir-contracts function filter.

charlielye committed:
attach to container in test.sh to catch error

charlielye committed:
reqres flakes.

charlielye committed:
crs stuff

charlielye committed:
instance name

charlielye committed:
yarn-project precommit hook

charlielye committed:
Merge branch 'cl/ci3-fake-base' into cl/ci3

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

charlielye committed:
fix yarn.lock. color precommit

charlielye committed:
precommit comment

charlielye committed:
fix

charlielye committed:
node crs read requested amount

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

charlielye committed:
jest silent reporter. NO_SPOT var.

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

charlielye committed:
Revert "feat: unified create circuit from acir (#10440)"

This reverts commit a4dfe13c1c0af3d527f5c9b2fcc38fe059e9bc38.

charlielye committed:
unbound fix

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

charlielye committed:
Reapply "feat: unified create circuit from acir (#10440)"

This reverts commit 4f54ca974125efe08b113c648f65c5d6458126f2.

charlielye committed:
ivc fix

charlielye committed:
better err handling with dump_fail

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

charlielye committed:
fix merge ci.yml

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

charlielye committed:
keys are not base64 encoded, just hex. because of course.

charlielye committed:
reverse hex for fields

charlielye committed:
disable assertions in bb for now, as leads to avm failures. remove flake on couple tests.

charlielye committed:
install xxd. terminate use iid.

charlielye committed:
fix

charlielye committed:
fix

charlielye committed:
region fix.

charlielye committed:
tweak

charlielye committed:
wip

charlielye committed:
try tmpfs

charlielye committed:
change hash

charlielye committed:
change hash

charlielye committed:
/tmp 1gb in-mem mount to spare disk io and start implementing limits

charlielye committed:
cpp fast formatting in test. ability to run just e2e flakes with test-e2e-flakes.

charlielye committed:
formatting

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

charlielye committed:
limit noir-projects to 16 jobs for mem reasons. merge fix.

charlielye committed:
Merge branch 'master' into cl/ci3

charlielye committed:
Merge branch 'master' into cl/ci3

charlielye committed:
revert test timings.

charlielye committed:
align acir tests with master

charlielye committed:
clean

charlielye committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

ludamad committed:
chore(ci): towards ci3 (#10751)

CI3 is a conceptual goal for uniting the CI flow and the dev flow as
much as possible, adding more depth to the bootstrap and build scripts
to be able to handle our needs.

This PR introduces all the work on CI3 so far, but still has an earthly
caller shell to make sure we can minimize the number of variables that
have changed at once.

There is a lot of changes in this PR.
See https://github.com/AztecProtocol/aztec-packages/pull/10711/files for
a subset of the changes without yarn.lock etc noise.

The big picture:
- The CI build has been made much less stateful. ci.yml now uses the ci3
bootstrap pattern, without fully moving off the earthly targets just
yet.
- The S3 cache mechanism is now the main cache mechanism. Note there is
no persistent disk now supporting the build.
There is a global cache on S3, readable without auth, that caches them
for 10 days. We no longer think of the build in terms of docker/buildkit
layers but instead as chunks that have different rebuild patterns that
match files in the monorepo.
- Moving to yarn 4.5.2. 

Niceties:
- faster builds due to script improvements and distributed cache
uploading by default
- work is more properly isolated in chunks from the above effort
- spot recovery is implemented, retrying with on-demand
- we no longer use github runners, side-stepping lots of edge-cases, and
instead rely on our builder realizing there is no work to do / hitting a
timeout via shutdown -P
- Docker images are no longer copied from the builder, meaning a large
class of flake is gone.

Non-niceties:
- The earthly setup is much less granular. There is two stages that have
their own one-layer builds. The earthly cache is fairly redundant, using
the S3 cache for most meaningful caching. (earthly will not be used in
ci.yml in the future)
- Some CI files are now duplicated, we will do a follow-on pass to get
rid of earthly helpers, build-system, etc
- CI currently also downloads the CI image fresh each time, will change
- we are currently pushing images to dockerhub with no expiration,
should move to ECR
- noir-projects currently retries once in the Earthfile as a last minute
issue was hit, will be fixed in a follow-up
- Issues: https://github.com/AztecProtocol/aztec-packages/issues/10775

---- 
WORKFLOW AFTER THIS PR:
- Run ./bootstrap.sh in root to bootstrap with cache, ./bootstrap.sh
full otherwise
- Run earthly +ci in root to 
- Put ci3 in your cache and note the ways to interact with ci in that
folder
- Note the new commands in ./bootstrap.sh like test-kind-network
- In yarn-project to run a single e2e test now use `test:e2e`. `test`
just runs the unit tests as per other projects.

---------

Co-authored-by: MirandaWood <miranda@aztecprotocol.com>
Co-authored-by: Charlie Lye <karl.lye@gmail.com>
Co-authored-by: Tom French <15848336+TomAFrench@users.noreply.github.com>

ludamad committed:
earthly syntax fix

ludamad committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

ludamad committed:
fix

ludamad committed:
robustness

ludamad committed:
more sane build-pic prep

ludamad committed:
more sane build-pic prep

ludamad committed:
fix

ludamad committed:
fix

ludamad committed:
better startup script

ludamad committed:
generate

ludamad committed:
Yarn

ludamad committed:
fix

ludamad committed:
fix

ludamad committed:
ensure-tester

ludamad committed:
files

ludamad committed:
files

ludamad committed:
S3_FORCE_UPLOAD

ludamad committed:
ensure-builder fix

ludamad committed:
start

ludamad committed:
fix

ludamad committed:
gogo

ludamad committed:
fix

ludamad committed:
fix

ludamad committed:
fix

ludamad committed:
fix

ludamad committed:
fix

ludamad committed:
gogo gadget bash error

ludamad committed:
lets go

ludamad committed:
fixes

ludamad committed:
input

ludamad committed:
input

ludamad committed:
update

ludamad committed:
update

ludamad committed:
update

ludamad committed:
update

ludamad committed:
update

ludamad committed:
update

ludamad committed:
run on tester lets go

ludamad committed:
try again

ludamad committed:
update

ludamad committed:
Update

ludamad committed:
fix

ludamad committed:
Updates

ludamad committed:
less output

ludamad committed:
less output

ludamad committed:
Update Earthfile

ludamad committed:
scripty stuff

ludamad committed:
Merge remote-tracking branch 'origin/cl/ci3-fixes' into cl/ci3-fixes

ludamad committed:
fix

ludamad committed:
fix

ludamad committed:
Update ci-arm.yml

ludamad committed:
fix

ludamad committed:
Merge remote-tracking branch 'origin/cl/ci3-fixes' into cl/ci3-fixes

ludamad committed:
the wrapped run command

ludamad committed:
uploaded

ludamad committed:
uploaded

ludamad committed:
update

ludamad committed:
fix

ludamad committed:
fix acir bench syntax

ludamad committed:
dont acir bench on e2e

ludamad committed:
less image time

ludamad committed:
dont run flakey e2e

ludamad committed:
syntax

ludamad committed:
parallel fix

charlielye committed:
cpp full bootstrap deletes build dirs first.

ludamad committed:
fill out set -eu

ludamad committed:
bootstrap -set eu

ludamad committed:
Make world state fail if no napi

ludamad committed:
try repair build

ludamad committed:
fix run_acir_tests.sh

ludamad committed:
Merge branch 'cl/ci3' into cl/ci3-fixes

ludamad committed:
update

ludamad committed:
syntax fix

ludamad committed:
fix bench summary

ludamad committed:
start kind full test sooner

ludamad committed:
Merge remote-tracking branch 'origin/master' into cl/ci3-fixes

ludamad committed:
run network test with yarn test

ludamad committed:
fast filter base images

ludamad committed:
Undo skip

ludamad committed:
simplify yarn hash, add cache_ls helper

ludamad committed:
just upload napi

ludamad committed:
ensure world state is uploaded

ludamad committed:
Fix kludge

ludamad committed:
state

ludamad committed:
Fix job name

ludamad committed:
Fix cache delete

ludamad committed:
try again

ludamad committed:
now make it pass the cache hoop

ludamad committed:
lets go

ludamad committed:
fix install parallel

ludamad committed:
update

ludamad committed:
less restrictive ci.sh

ludamad committed:
for build

ludamad committed:
fix ci.sh

ludamad committed:
try downsizing testers

ludamad committed:
try e2e with standard github runner

ludamad committed:
try again

ludamad committed:
denoise

ludamad committed:
update

ludamad committed:
try no exit spot

ludamad committed:
retries

ludamad committed:
retries

ludamad committed:
fix

ludamad committed:
fix

ludamad committed:
attach ebs

ludamad committed:
update

ludamad committed:
resinstate maybe exit spot

ludamad committed:
fix yarn project cache

ludamad committed:
bootstrap

ludamad committed:
update

ludamad committed:
cache fix

ludamad committed:
install kludge

ludamad committed:
fix

ludamad committed:
remove redundant 4epochs test

ludamad committed:
simplify

ludamad committed:
fix

ludamad committed:
simplify more

ludamad committed:
lets go

ludamad committed:
simpler transfer tests

ludamad committed:
pare down network tests

ludamad committed:
less earthly nag

ludamad committed:
Update

ludamad committed:
hash

ludamad committed:
try it out

ludamad committed:
more conditionality

ludamad committed:
edge ci

ludamad committed:
refactor vm tests

ludamad committed:
Docs selection of ci

ludamad committed:
dont make GH runner token mandatory

ludamad committed:
simplify workflows

ludamad committed:
Merge remote-tracking branch 'origin/master' into cl/ci3-fixes

ludamad committed:
-

ludamad committed:
update prover client fix

ludamad committed:
simplify docs

ludamad committed:
quieter network

ludamad committed:
fix nargo

ludamad committed:
bring back earthly retries

ludamad committed:
sufferage

ludamad committed:
lets try again

ludamad committed:
flakefest

ludamad committed:
use ref name for arm concurrency

ludamad committed:
lets try it again

ludamad committed:
woopsy

ludamad committed:
arm fixes

ludamad committed:
changes

ludamad committed:
fix

ludamad committed:
fix bench e2e

ludamad committed:
Merge remote-tracking branch 'origin/master' into cl/ci3

ludamad committed:
Update ci.yml
